Unpickling WCS object raises warning

### Description
When a WCS object is unpickled, a warning is raised:
```
>>> pickle.loads(pickle.dumps(wcs))
WARNING: FITSFixedWarning: The WCS transformation has more axes (2) than the image it is associated with (0) [astropy.wcs.wcs]
```

I've dug into it a bit, though I don't have the expertise to know for sure what's happening. The warning appears to be raised [here](https://github.com/astropy/astropy/blob/main/astropy/wcs/wcs.py#L519), based on a comparison of `wcsprm.naxis` and `header.get('NAXIS', None)`. The WCS class supports pickling by implementing the `__reduce__()` method, which serializes the WCS object by calling its `to_fits()` method, which in turns calls `to_header()`. The documentation of `to_header()` makes clear it does not include an `NAXIS` key in the generated header. `to_fits()` passes the header into a new `PrimaryHDU`, and the header of this generated `PrimaryHDU` has the `NAXIS` key present with a value of 0. The `HDUList` generated by `to_fits()` is returned by `__reduce__()`. The unpickling goes through the `__WCS_unpickle__` function, which extracts the header from that hdulist and uses it to create a new WCS. `header.get('NAXIS', None)` therefore is 0, the value that showed up during `to_fits()`, while `wcsprm` must detect some other way that there are two axes.

Within `to_fits()`, changing the line

```
        hdu = fits.PrimaryHDU(header=header)
```
to
```
        hdu = fits.PrimaryHDU(header=header, data=np.empty([0]*self.naxis))
```
eliminates the warning (by ensuring a correct value is detected when `NAXIS` is added to the header, I suppose), but I don't know if that change could be problematic in some other way. That change does cause six tests in `wcs/tests/test_pickle.py` to fail. Five of them appear to be failing because they expect this warning to be raised (e.g. `FAILED wcs/tests/test_pickle.py::test_basic - Failed: DID NOT WARN. No warnings of type (<class 'astropy.wcs.wcs.FITSFixedWarning'>,) were emitted. The list of emitted warnings is: [].`), and I'm afraid I don't understand why the sixth fails (`FAILED wcs/tests/test_pickle.py::test_subclass - AttributeError: 'Sub' object has no attribute 'naxis'`).

The fact that some tests seem to require this warning to be raised makes me wonder if this is intended behavior. But the fact that `NAXIS=0` is being inserted within `to_fits()` feels like it might be a default value that's inappropriate for this case, and I think it seems reasonable to expect to pickle and unpickle `WCS` objects without warnings.

### Expected behavior
Unpickling `WCS` objects doesn't raise errors

### Actual behavior
```
In [1]: from astropy.wcs import WCS

In [2]: import pickle

In [3]: wcs = WCS(naxis=2)

In [4]: pickled = pickle.dumps(wcs)

In [5]: pickle.loads(pickled)
WARNING: FITSFixedWarning: The WCS transformation has more axes (2) than the image it is associated with (0) [astropy.wcs.wcs]
Out[5]: 
WCS Keywords

Number of WCS axes: 2
CTYPE : ''  ''  
CRVAL : 0.0  0.0  
CRPIX : 0.0  0.0  
PC1_1 PC1_2  : 1.0  0.0  
PC2_1 PC2_2  : 0.0  1.0  
CDELT : 1.0  1.0  
NAXIS : 0  0

In [6]: 
```


### System Details
```
>>> import platform; print(platform.platform())
macOS-12.2-arm64-arm-64bit
>>> import sys; print("Python", sys.version)
Python 3.10.2 | packaged by conda-forge | (main, Feb  1 2022, 19:29:01) [Clang 11.1.0 ]
>>> import numpy; print("Numpy", numpy.__version__)
Numpy 1.22.2
>>> import erfa; print("pyerfa", erfa.__version__)
pyerfa 2.0.0.1
>>> import astropy; print("astropy", astropy.__version__)
astropy 5.1.dev551+g314dbe860
>>> import scipy; print("Scipy", scipy.__version__)
Scipy 1.8.0
>>> import matplotlib; print("Matplotlib", matplotlib.__version__)
Matplotlib 3.5.1
```

--- a/astropy/wcs/tests/test_pickle.py
+++ b/astropy/wcs/tests/test_pickle.py
@@ -19,8 +19,7 @@ from astropy.wcs.wcs import FITSFixedWarning
 def test_basic():
     wcs1 = wcs.WCS()
     s = pickle.dumps(wcs1)
-    with pytest.warns(FITSFixedWarning):
-        pickle.loads(s)
+    pickle.loads(s)
 
 
 def test_dist():
@@ -33,8 +32,7 @@ def test_dist():
         assert wcs1.det2im2 is not None
 
         s = pickle.dumps(wcs1)
-        with pytest.warns(FITSFixedWarning):
-            wcs2 = pickle.loads(s)
+        wcs2 = pickle.loads(s)
 
         with NumpyRNGContext(123456789):
             x = np.random.rand(2 ** 16, wcs1.wcs.naxis)
@@ -52,8 +50,7 @@ def test_sip():
             wcs1 = wcs.WCS(hdulist[0].header)
         assert wcs1.sip is not None
         s = pickle.dumps(wcs1)
-        with pytest.warns(FITSFixedWarning):
-            wcs2 = pickle.loads(s)
+        wcs2 = pickle.loads(s)
 
         with NumpyRNGContext(123456789):
             x = np.random.rand(2 ** 16, wcs1.wcs.naxis)
@@ -71,8 +68,7 @@ def test_sip2():
             wcs1 = wcs.WCS(hdulist[0].header)
         assert wcs1.sip is not None
         s = pickle.dumps(wcs1)
-        with pytest.warns(FITSFixedWarning):
-            wcs2 = pickle.loads(s)
+        wcs2 = pickle.loads(s)
 
         with NumpyRNGContext(123456789):
             x = np.random.rand(2 ** 16, wcs1.wcs.naxis)
@@ -90,8 +86,7 @@ def test_wcs():
 
     wcs1 = wcs.WCS(header)
     s = pickle.dumps(wcs1)
-    with pytest.warns(FITSFixedWarning):
-        wcs2 = pickle.loads(s)
+    wcs2 = pickle.loads(s)
 
     with NumpyRNGContext(123456789):
         x = np.random.rand(2 ** 16, wcs1.wcs.naxis)
@@ -103,16 +98,35 @@ def test_wcs():
 
 class Sub(wcs.WCS):
     def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
         self.foo = 42
 
 
 def test_subclass():
-    wcs = Sub()
-    s = pickle.dumps(wcs)
-    with pytest.warns(FITSFixedWarning):
-        wcs2 = pickle.loads(s)
+    wcs1 = Sub()
+    wcs1.foo = 45
+    s = pickle.dumps(wcs1)
+    wcs2 = pickle.loads(s)
 
     assert isinstance(wcs2, Sub)
-    assert wcs.foo == 42
-    assert wcs2.foo == 42
+    assert wcs1.foo == 45
+    assert wcs2.foo == 45
     assert wcs2.wcs is not None
+
+
+def test_axes_info():
+    w = wcs.WCS(naxis=3)
+    w.pixel_shape = [100, 200, 300]
+    w.pixel_bounds = ((11, 22), (33, 45), (55, 67))
+    w.extra = 111
+
+    w2 = pickle.loads(pickle.dumps(w))
+
+    # explicitly test naxis-related info
+    assert w.naxis == w2.naxis
+    assert w.pixel_shape == w2.pixel_shape
+    assert w.pixel_bounds == w2.pixel_bounds
+
+    # test all attributes
+    for k, v in w.__dict__.items():
+        assert getattr(w2, k) == v

--- a/astropy/wcs/tests/test_pickle.py
+++ b/astropy/wcs/tests/test_pickle.py
@@ -7,7 +7,8 @@ import numpy as np
 import pytest
 from numpy.testing import assert_array_almost_equal
 
-from astropy.utils.data import get_pkg_data_contents, get_pkg_data_fileobj
+from astropy.utils.data import (get_pkg_data_contents, get_pkg_data_fileobj,
+                                get_pkg_data_filename)
 from astropy.utils.exceptions import AstropyDeprecationWarning
 from astropy.utils.misc import NumpyRNGContext
 from astropy.io import fits
@@ -130,3 +131,30 @@ def test_axes_info():
     # test all attributes
     for k, v in w.__dict__.items():
         assert getattr(w2, k) == v
+
+
+def test_pixlist_wcs_colsel():
+    """
+    Test selection of a specific pixel list WCS using ``colsel``. See #11412.
+    """
+    hdr_file = get_pkg_data_filename('data/chandra-pixlist-wcs.hdr')
+    hdr = fits.Header.fromtextfile(hdr_file)
+    with pytest.warns(wcs.FITSFixedWarning):
+        w0 = wcs.WCS(hdr, keysel=['image', 'pixel'], colsel=[11, 12])
+
+    with pytest.warns(wcs.FITSFixedWarning):
+        w = pickle.loads(pickle.dumps(w0))
+
+    assert w.naxis == 2
+    assert list(w.wcs.ctype) == ['RA---TAN', 'DEC--TAN']
+    assert np.allclose(w.wcs.crval, [229.38051931869, -58.81108068885])
+    assert np.allclose(w.wcs.pc, [[1, 0], [0, 1]])
+    assert np.allclose(w.wcs.cdelt, [-0.00013666666666666, 0.00013666666666666])
+    assert np.allclose(w.wcs.lonpole, 180.)
+
+
+def test_alt_wcskey():
+    w = wcs.WCS(key='A')
+    w2 = pickle.loads(pickle.dumps(w))
+
+    assert w2.wcs.alt == 'A'

--- a/astropy/wcs/wcs.py
+++ b/astropy/wcs/wcs.py
@@ -382,6 +382,12 @@ class WCS(FITSWCSAPIMixin, WCSBase):
                  fix=True, translate_units='', _do_set=True):
         close_fds = []
 
+        # these parameters are stored to be used when unpickling a WCS object:
+        self._init_kwargs = {
+            'keysel': copy.copy(keysel),
+            'colsel': copy.copy(colsel),
+        }
+
         if header is None:
             if naxis is None:
                 naxis = 2
@@ -2994,8 +3000,11 @@ reduce these to 2 dimensions using the naxis kwarg.
         buffer = io.BytesIO()
         hdulist.writeto(buffer)
 
+        dct = self.__dict__.copy()
+        dct['_alt_wcskey'] = self.wcs.alt
+
         return (__WCS_unpickle__,
-                (self.__class__, self.__dict__, buffer.getvalue(),))
+                (self.__class__, dct, buffer.getvalue(),))
 
     def dropaxis(self, dropax):
         """
@@ -3288,9 +3297,11 @@ def __WCS_unpickle__(cls, dct, fits_data):
         for k, na in enumerate(naxes):
             hdulist[0].header[f'naxis{k + 1:d}'] = na
 
+    kwargs = dct.pop('_init_kwargs', {})
     self.__dict__.update(dct)
 
-    WCS.__init__(self, hdulist[0].header, hdulist)
+    wcskey = dct.pop('_alt_wcskey', ' ')
+    WCS.__init__(self, hdulist[0].header, hdulist, key=wcskey, **kwargs)
     self.pixel_bounds = dct.get('_pixel_bounds', None)
 
     return self

--- a/astropy/wcs/wcs.py
+++ b/astropy/wcs/wcs.py
@@ -3277,12 +3277,21 @@ def __WCS_unpickle__(cls, dct, fits_data):
     """
 
     self = cls.__new__(cls)
-    self.__dict__.update(dct)
 
     buffer = io.BytesIO(fits_data)
     hdulist = fits.open(buffer)
 
+    naxis = dct.pop('naxis', None)
+    if naxis:
+        hdulist[0].header['naxis'] = naxis
+        naxes = dct.pop('_naxis', [])
+        for k, na in enumerate(naxes):
+            hdulist[0].header[f'naxis{k + 1:d}'] = na
+
+    self.__dict__.update(dct)
+
     WCS.__init__(self, hdulist[0].header, hdulist)
+    self.pixel_bounds = dct.get('_pixel_bounds', None)
 
     return self
 

--- a/docs/changes/wcs/12844.bugfix.rst
+++ b/docs/changes/wcs/12844.bugfix.rst
@@ -1,0 +1,5 @@
+Fixed a bug due to which ``naxis``, ``pixel_shape``, and
+``pixel_bounds`` attributes of ``astropy.wcs.WCS`` were not restored when
+an ``astropy.wcs.WCS`` object was unpickled. This fix also eliminates
+``FITSFixedWarning`` warning issued during unpiclikng of the WCS objects
+related to the number of axes.

--- a/docs/changes/wcs/12844.bugfix.rst
+++ b/docs/changes/wcs/12844.bugfix.rst
@@ -2,4 +2,6 @@ Fixed a bug due to which ``naxis``, ``pixel_shape``, and
 ``pixel_bounds`` attributes of ``astropy.wcs.WCS`` were not restored when
 an ``astropy.wcs.WCS`` object was unpickled. This fix also eliminates
 ``FITSFixedWarning`` warning issued during unpiclikng of the WCS objects
-related to the number of axes.
+related to the number of axes. This fix also eliminates errors when
+unpickling WCS objects originally created using non-default values for
+`key`, `colsel`, and `keysel` parameters.

